### PR TITLE
fix #19 and #25 : splited core.py and test_la.py

### LINKED ISSUE #19: split core.py
* at a minimum all Axis related stuff should go to an "axis" module
Axis, AxisCollection
* potential other modules:
group
expr

### LINKED ISSUE #25: split unit tests
at a minimum, move Axis, AxisCollection and LGroup tests out of test_la.py

### PR DESCRIPTION
**Pull request by alixdamman**
fix #19 : splited core.py in expr.py, group.py, axis.py and array.py

fix #25 : splited test_la.py in common.py, test_axis.py, test_group.py and test_array.py

+ reorganized packages:
- core: expr.py, group.py, axis.py, array.py, session.py, ufuncs.py, ipfp.py
- io: excel.py
- util: misc.py, oset.py
- viewer: api.py, model.py, view.py
- tests: common.py, test_array.py, test_axis.py, test_group.py, test_session.py, test_excel.py, test_ipfp.py

### CODE REVIEW

File: larray/core/array.py [hunk 1]
```diff
@@ -111,13 +101,11 @@
...
-from larray.oset import *
-from larray.utils import (table2str, size2str, unique, csv_open, long,
-                          decode, basestring, unicode, bytes, izip, rproduct,
-                          ReprString, duplicates, array_lookup2, strip_rows,
-                          skip_comment_cells, find_closing_chr, StringIO, PY2,
-                          float_error_handler_factory)
-
+from larray.util.utils import (table2str, size2str, unique, decode, basestring, izip, rproduct, ReprString, StringIO,
+                               duplicates, float_error_handler_factory)
+from .axis_group import (Axis, AxisCollection, x, Group, PGroup, LGroup, ExprNode, remove_nested_groups, _isnoneslice,
```
Comment (gdementen) [inline larray/core/array.py]: I would rather use absolute import like other lines (or convert them all to relative)

File: larray/core/array.py [hunk 2]
```diff
@@ -10388,23 +6734,16 @@ def read_excel(filepath, sheetname=0, nb_index=None, index_col=None, na=np.nan,
...
                             .format(list(kwargs.keys())[0]))
-        from .excel import open_excel
+        from larray.io.excel import open_excel
         with open_excel(filepath) as wb:
-            return wb[sheetname].load(index_col=index_col)
+            return wb[sheetname].load(index_col=_get_index_col(nb_index, index_col))
     else:
-        df = pd.read_excel(filepath, sheetname, index_col=index_col, engine=engine, **kwargs)
-        return df_aslarray(df, sort_rows=sort_rows, sort_columns=sort_columns, raw=index_col is None, fill_value=na)
+        df, raw = _read_excel(filepath, sheetname, nb_index, index_col, engine, **kwargs)
```
Comment (gdementen) [inline larray/core/array.py]: this PR is already hard enough to review as it is, could you move this change to a separate PR (possibly depending on this one)

File: larray/core/axis_group.py [hunk 3]
```diff
@@ -0,0 +1,3635 @@
...
+    >>> index_by_id([age, sex, time], gender)
+    1
+    """
+    for i, item in enumerate(seq):
+        if item is value:
+            return i
+    raise ValueError("%s is not in list" % value)
+
+
+class ExprNode(object):
```
Comment (gdementen) [inline larray/core/axis_group.py]: we might want to move this class and related code to its own module

File: larray/core/axis_group.py [hunk 4]
```diff
@@ -0,0 +1,3635 @@
...
+
+
+def _isnoneslice(v):
+    """
+    Checks if input is slice(None) object.
+    """
+    return isinstance(v, slice) and v.start is None and v.stop is None and v.step is None
+
+
+def _seq_summary(seq, n=3, repr_func=repr, sep=' '):
```
Comment (gdementen) [inline larray/core/axis_group.py]: this could move to utils

File: larray/core/axis_group.py [hunk 5]
```diff
@@ -0,0 +1,3635 @@
...
+    '0 1 ... 8 9'
+    """
+    if len(seq) <= 2 * n:
+        short_seq = [repr_func(v) for v in seq]
+    else:
+        short_seq = [repr_func(v) for v in seq[:n]] + ['...'] + [repr_func(v) for v in seq[-n:]]
+    return sep.join(short_seq)
+
+
+def index_by_id(seq, value):
```
Comment (gdementen) [inline larray/core/axis_group.py]: move this to utils?

File: larray/core/axis_group.py [hunk 6]
```diff
@@ -0,0 +1,3635 @@
...
+    >>> union('a', 'a, b, c, d', ['d', 'e', 'f'], '..2')
+    ['a', 'b', 'c', 'd', 'e', 'f', 0, 1, 2]
+    """
+    if args:
+        return list(unique(chain(*(_to_ticks(arg) for arg in args))))
+    else:
+        return []
+
+
+def _isnoneslice(v):
```
Comment (gdementen) [inline larray/core/axis_group.py]: move this to utils?

File: larray/io/session_io.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
+# -*- coding: utf8 -*-
+from __future__ import absolute_import, division, print_function
+
+import os
+from pandas import ExcelWriter, ExcelFile, HDFStore
+from collections import OrderedDict
+
+from .excel import open_excel
```
Comment (gdementen) [inline larray/io/session_io.py]: change to absolute import?

File: larray/tests/test_group.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
+from __future__ import absolute_import, division, print_function
+
+from unittest import TestCase
+
+import pytest
+import numpy as np
+
+from .common import abspath, assert_array_equal, assert_array_nan_equal
```
Comment (gdementen) [inline larray/tests/test_group.py]: absolute import?

File: larray/tests/test_la.py [hunk 9]
```diff
@@ -13,64 +12,13 @@
 except ImportError:
     xw = None
 
-from larray import (LArray, Axis, AxisCollection, LGroup, LSet, PGroup, union,
+from .common import abspath, assert_array_equal, assert_array_nan_equal
```
Comment (gdementen) [inline larray/tests/test_la.py]: absolute import?

File: larray/tests/test_la.py [hunk 10]
```diff
@@ -3429,6 +2393,7 @@ def test_read_excel_pandas(self):
         assert_array_equal(la, [3722, 3395, 3347])
 
         la = read_excel(abspath('test.xlsx'), '2d', nb_index=1, engine='xlrd')
+        print('la\n', la)
```
Comment (gdementen) [inline larray/tests/test_la.py]: kill line?

**Comment by alixdamman** [comment]
splited core.py in expr.py, group.py, axis.py and array.py + removed io.py (and array_io.py and session_io.py)

File: larray/core/array.py [hunk 11]
```diff
@@ -10285,12 +6621,7 @@ def read_csv(filepath_or_buffer, nb_index=None, index_col=None, sep=',', headers
         kwargs['header'] = 1
         kwargs['comment'] = '#'
 
-    if nb_index is not None and index_col is not None:
-        raise ValueError("cannot specify both nb_index and index_col")
-    elif nb_index is not None:
-        index_col = list(range(nb_index))
-    elif isinstance(index_col, int):
-        index_col = [index_col]
+    index_col = _get_index_col(nb_index, index_col)
```
Comment (gdementen) [inline larray/core/array.py]: please move this change (factorization of _get_index_col) to another PR. It is a good change, but such a large PR should only contain what is strictly necessary, otherwise extra stuff like this get completely lost/can go unnoticed.

File: larray/core/axis.py [hunk 12]
```diff
@@ -0,0 +1,2209 @@
...
+
+import re
+import sys
+import warnings
+from itertools import product
+
+import numpy as np
+
+from larray.core.expr import ExprNode
+from larray.core.group import (ABCAxis, ABCAxisReference, ABCLArray, Group, LGroup, PGroup, PGroupMaker,
```
Comment (gdementen) [inline larray/core/axis.py]: those 3 ABC should be in their own "abc" module I think. Having them in group.py seems completely arbitrary.

File: larray/ipfp/__init__.py [hunk 13]
```diff
@@ -0,0 +1,6 @@
+from __future__ import absolute_import
+
+import warnings
+from larray.ipfp.ipfp import *
+
+warnings.warn("ipfp function should imported as\nfrom larray import *\nor\nimport larray.ipfp", DeprecationWarning)
```
Comment (gdementen) [inline larray/ipfp/__init__.py]: nor -> not
and 
from larray import * -> from larray import ipfp (or not imported explicitly at all if you use "from larray import *")

File: larray/core/convert.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
+from __future__ import absolute_import, print_function
```
Comment (gdementen) [inline larray/core/convert.py]: is there a reason this module (or the functions from this module) are not somewhere in the io package?

Comment (alixdamman) [inline larray/core/convert.py]: I consider that df_asslarray is not an "IO" function. But OK, I can let df_aslarray in array.py

**Comment by gdementen** [comment]
LGTM, Thanks
